Add external authorization guide (#144)
openservicemesh/osm-docs: content/docs/tasks/external_auth_opa.md

# External Authorization

> ## Overview and limitations

Author: I'll split them.

External authorization allows offloading authorization of incoming HTTP requests to a remote endpoint not necessarily running inside the proxy's context.
Proxies supporting external authorization will, for every request, issue a permission check to the remote endpoint. Due to the potential high number of RPS a system can have, external authorization has a number of tuneable settings to allow more or less control at the expense of performance (send HTTP payload or headers only, timeout, default behaviour if authorization fails to reply, etc.).

> OSM allows configuring envoy's [External Authorization extension](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter) through OSM's MeshConfig.

Author: yep. will do.

Authorization filter is currently applied to inmesh `inbound` and `ingress` connections affecting HTTP request made towards any service in the mesh.

The following example will show how to configure external authorization with `opa-envoy-plugin`.
Note this example will use a single, remote (over the network) endpoint to validate all traffic. This is not a suggested production deployment.

> Example OSM's MeshConfig:

Author: I reworked this a bit. Let's see if now makes a bit more sense.

> Example OSM's MeshConfig:

Author: @zr-msft I'm not sure I understood you here. On markdown, h2 seems like huge for individual steps. I'll push again latest changes and maybe discuss over it as I addressed quite a bit of it.

```
...
inboundExternalAuthorization:
  enable: true
  address: opa.opa.svc.cluster.local
  port: 9191
  statPrefix: authz_opa
  timeout: 1s
  failureModeAllow: false
```

> ## Example Walkthrough

Author: renamed section to `OSM with OPA plugin external authorization walkthrough`, lmk

> - Deploy an [opa-envoy-plugin](https://github.com/open-policy-agent/opa-envoy-plugin), you will need to back the service with network connectivity for this walkthrough - we highly suggest to read carefully `ola-envoy-plugin`'s guide first.

Author: this part and the specific details have been clarified, also pointing the a sample deployment that we added on OSM main branch. hopefully this is less vague and more clear and concise

```
kubectl apply -f https://raw.githubusercontent.com/open-policy-agent/opa-envoy-plugin/main/quick_start.yaml
```

> Once `opa-envoy-plugin` is running and reachable through the network, you can proceed to deploy OSM's Demo, follow `demo/run-osm-demo.sh`

Reviewer: ```suggestion
Once `opa-envoy-plugin` is running, verify you can access it through the network by ...

 ## Deploy the OSM demo
 Deploy the OSM's Demo using [demo/run-osm-demo.sh](PROVIDE LINK HERE):
```

If the user needs to clone a whole repo to run the demo, you need to tell them that first in this section

```

> demo/run-osm-demo.sh  # wait for all services to come up

Reviewer: ```suggestion
cd PATH_TO_DEMO
./run-osm-demo.sh 
```

Also tell them how to verify all the services are up. You can show them the sample output, or describe what it looks like.

Author: this is part of the https://github.com/openservicemesh/osm/tree/main/demo#how-to-run-the-osm-automated-demo. I'll point users to the demo documentation, I don't think we should be repeating steps in multiple articles

```

> Followingly, edit OSM's meshconfig to add external authorization to the mesh pods:

Reviewer: ```suggestion
## Add external authorization

Edit OSM's meshconfig to add external authorization to the mesh pods using `kubectl edit`:
```

```
kubectl edit meshconfig osm-mesh-config -n osm-system
...
inboundExternalAuthorization:
      enable: true
      address: opa.opa.svc.cluster.local  ## or reachable FQDN for opa plugin running
      port: 9191
      failureModeAllow: false
      statPrefix: inboundExtAuthz
      timeout: 1s

```

> Traffic should fail right out of the bat:

Reviewer: ```suggestion
Confirm you 403 errors when trying to reach the `bookstore` service.
```

```
kubectl logs <bookbuyer_pod> -n bookbuyer bookbuyer
```
```
...
--- bookbuyer:[ 8 ] -----------------------------------------

Fetching http://bookstore.bookstore:14001/books-bought
Request Headers: map[Client-App:[bookbuyer] User-Agent:[Go-http-client/1.1]]
Identity: n/a
Booksbought: n/a
Server: envoy
Date: Tue, 04 May 2021 01:20:39 GMT
Status: 403 Forbidden
ERROR: response code for "http://bookstore.bookstore:14001/books-bought" is 403;  expected 200
...
```

> You shoud also be able to see the logs in `opa-envoy-plugin` for rejected authorization calls:

Reviewer: ```suggestion
Also, verify the logs in `opa-envoy-plugin` show rejected authorization calls:
```

```
kubectl logs opa -n opa
...
{"decision_id":"1df154b5-658a-47bf-ac18-be52998605da"
...
"result":false,   // resulting decision
...
"time":"2021-05-04T01:21:18Z","timestamp":"2021-05-04T01:21:18.1971808Z","type":"openpolicyagent.org/decision_logs"
}
```

> - Now edit OPA's policy:

Reviewer: ```suggestion
## Update the OPA policy

Edit the OPA policy using `kubectl edit`:
- Now edit OPA's policy:
```

Author: let's discuss after addressing latest comments how to do this.

```
kubectl edit configmap opa-policy -n opa
```

> For the sake of testing end to end, look for a default rule such as:

Reviewer: ```suggestion
For this example, update the default allow rule:
```

```
default allow = false
```

> and change it to:

Reviewer: ```suggestion
Change the value to `true`:
```

```
default allow = true
```

> Finally, restart `opa-envoy-plugin`:

Reviewer: ```suggestion
Restart the `opa-envoy-plugin` using `kubectl rollout`:
```

```
kubectl rollout restart deployment opa -n opa
```

> - Observe that bookbuyer calls are now being allowed:

Reviewer: ```suggestion
Verify that requests to the `bookbuyer` service now being allowed:
```

```
--- bookbuyer:[ 2663 ] -----------------------------------------

Fetching http://bookstore.bookstore:14001/books-bought
Request Headers: map[Client-App:[bookbuyer] User-Agent:[Go-http-client/1.1]]
Identity: bookstore-v1
Booksbought: 1087
Server: envoy
Date: Tue, 04 May 2021 02:00:46 GMT
Status: 200 OK
MAESTRO! THIS TEST SUCCEEDED!

Fetching http://bookstore.bookstore:14001/buy-a-book/new
Request Headers: map[]
Identity: bookstore-v1
Booksbought: 1088
Server: envoy
Date: Tue, 04 May 2021 02:00:47 GMT
Status: 200 OK
ESC[90m2:00AMESC[0m ESC[32mINFESC[0m BooksCountV1=21490056 ESC[36mcomponent=ESC[0mdemo ESC[36mfile=ESC[0mbooks.go:167
MAESTRO! THIS TEST SUCCEEDED!
```

```
{"decision_id":"3f29d449-7f71-4721-b93c-ad7d375e0f80",
...
"result":true,
...
"time":"2021-05-04T02:01:35Z","timestamp":"2021-05-04T02:01:35.816768454Z","type":"openpolicyagent.org/decision_logs"}
```

> More complex authorization code can now be added on top of OPA's policy. Also important to note that with an external entity  operating through the network, a lot of latency is introduced to authorize every HTTP call. To mitigate this overhead, OPA has it's own injector that can inject policy agents on the same pod next to envoy, reducing the stack's latency to localhost communication between envoy and OPA plugin. For additional notes on their deployment model, check their [Open Policy Agent official documentation](https://www.openpolicyagent.org/docs/latest/).

Reviewer: ```suggestion
This example rule shows how you can use the OPA plugin to enforce authorization policies on network traffic and is not intended for production deployments. For example, enforcing authorization on every HTTP call in this manner adds significant overhead, which can be reduced using the OPA injector. For more information on using OPA, see [Open Policy Agent official documentation](https://www.openpolicyagent.org/docs/latest/).
```

---
title: "External Authorization"
description: "Configure external authorization through MeshConfig."
type: docs
---
# External Authorization

## Overview
External authorization allows offloading authorization of incoming HTTP requests to a remote endpoint not necessarily running inside the proxy's context.
Proxies supporting external authorization will, for every request, issue a permission check to the remote endpoint. Due to the potential high number of RPS a system can have, external authorization has a number of tuneable settings to allow more or less control at the expense of performance (send HTTP payload or headers only, timeout, default behaviour if authorization fails to reply, etc.).

OSM allows configuring envoy's [External Authorization extension](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter) through OSM's MeshConfig.

## Limitations

> Currently, authorization filtering can't be dinamically configured to be installed in subsets or groups of proxies defined by the user, and is instead globally applied for all services in the mesh when enabled.

Reviewer 2: > dinamically 

dynamically

Similarly, the filtering direction is to be statically applied to  `inbound` and `ingress` connections within the mesh, affecting any and all HTTP request made towards any service or application in the mesh when enabled.

## OSM with OPA plugin external authorization walkthrough

> The following section will document how to configure external authorization in conjuction with `opa-envoy-plugin`.

Reviewer 2: > conjuction

conjunction

We strongly suggest to go through [OPA envoy plugin's](https://github.com/open-policy-agent/opa-envoy-plugin) documentation first to further understand their configuration options and deployment models.

The following example uses a single, remote (over the network) endpoint to validate all traffic. This configuration is not recommended for a production deployment.

- First, start by deploying OSM's Demo. We will use this sample deployment to test external authorization capabilities. Please refer to [OSM's Automated Demo](https://github.com/openservicemesh/osm/tree/main/demo#how-to-run-the-osm-automated-demo) and follow the instructions.

```
# Assuming OSM repo is available
cd <PATH_TO_OSM_REPO>

> demo/run-osm-demo.sh  # wait for all services to come up

Reviewer 3: @eduser25 Since the website primarily relies on the manual demo and corresponding manifests, it might make more sense to show OPA integration with the manual demo as opposed to the automated demo.

Author: guess I can make peace by just pointing to the manual demo for the sake of not going through all the steps again?
The end result for each is not much different from both demos after all I guess

Author: I'll leave it as is for now and leave an open issue on me to move it properly to manual when able.

Reviewer 3: @eduser25 please open an issue to track this.
